articles/ros2_dds_security_pkcs11_support.md

```diff
@@ -0,0 +1,86 @@
+---
+layout: default
+title: ROS2 DDS security PKCS#11 URI support
+abstract:
+  The [DDS-Security specification][dds_security] defines the use of Hardware Security Modules (HSM) and PKCS#11 URIs as an alternative to private keys and certificates stored in the file system.
+  Current implementation only supports these tokens to be directly stored in the file system as `.pem` files.
+  This is a design proposal to support PKCS#11 URIs.
+author: '[Iker Luengo](https://github.com/IkerLuengo)'
+published: false
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Scope
+
+The [DDS-Security specification][dds_security] requires the security documents (private keys, certificates, governance and permissions) to be provided to the DDS implementation within the participant's `properties`.
+It defines a set of reserved properties, that must hold the URI of the corresponding document.
+However, it allows different URI schemes to be used.
+Specifically, in the case of certificates and private keys, it defines the support for:
+
+- file scheme (`file:/keystore/enclaves/foo/key.pem`).
+- PKCS#11 scheme (`pkcs11:object=MyParticipantPrivateKey;type=private`).
+- data scheme (`data:,-----BEGIN RSA PRIVATE KEY----- MIIEpAIBAAKCAQEA3HIh...AOBaaqSV37XBUJg== -----END RSA PRIVATE KEY-----`).
+
+Current RCL impementation only support the `file` scheme.
+Furthermore, it searches these security files in an enclave subdirectory within the reserved `enclaves` subfolder of the root keystore, corresponding to the fully-qualified path of every enclave.
+For example, for the `/foo/bar` enclave, the directory structure would look like:
+
+    <root>
+    ├── enclaves
+    │   └── foo
+    │       └── bar
+    │           ├── cert.pem
+    │           ├── key.pem
+    │           ├── ...
+    └── public
+        ├── ...
+
+
+Note that it also requires the names of the files to be the ones expected.
+Specifically, all certificate and key files must have the `.pem` extension.
+In order to configure the security properties of a DDS participant, the path of the appropriate file in the enclave directory is added as the `file` URI of the corresponding property.
+For example, for the private key in the authentication plugin:
+
+    dds.sec.auth.private_key = file:<root>/enclaves/foo/bar/key.pem
+
+## Proposal
+
+### Goals
+
+Support PKCS#11 URIs for certificates and key files.
+`data` URIs are out of scope of this proposal.
+
+### Specification
+
+We want to keep the current keystore structure as much as possible, as this will enable to keep all the current implementation regarding the enclave management and the CLI features that help setting up the keystore.
```

Comment:
```suggestion
We want to keep the current keystore structure as much as possible, as this will enable to keep all the current implementations regarding the enclave management and the CLI features that help setting up the keystore.
```